source/client-side-encryption/client-side-encryption.rst

```diff
@@ -597,6 +631,66 @@ message.
 
 Return "access_token" as the credential.
 
+
+Obtaining an Access Token for Azure Key Vault
+`````````````````````````````````````````````
+
+.. versionadded:: 1.12.0 2022/08/16
+
+Virtual machines running on the Azure platform have one or more *Managed
+Identities* associated with them. From within the VM, an identity can be used by
+obtaining a access token via HTTP from the *Azure Instance Metadata Service*
+(IMDS). `See this documentation for more information`__
+
+__ https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-to-use-vm-token#get-a-token-using-http
+
+.. default-role:: math
+
+If an Azure VM has more than one managed identity, requesting an access token
+requires additional query parameters |--| `objectId`, `clientId`, and `miResId`
+|--| that disambiguate the request. We'll refer to these parameters as `M`. If
+the VM only has one managed identity, `M` may be omitted.
+
```

Comment:
Should the driver cache the returned `access_token`?

https://docs.microsoft.com/en-us/azure/virtual-machines/windows/instance-metadata-service?tabs=windows#rate-limiting suggests there are rate limits:

> In general, requests to IMDS are limited to 5 requests per second (on a per VM basis)

This was a problem for AWS and is being resolved in https://jira.mongodb.org/browse/DRIVERS-2333.

This was not a problem for GCP, since the GCP Metadata Server does not have a rate limit. The GCP Metadata Server caches access tokens. See: https://docs.google.com/document/d/1RTDp5QMg_ayYnR_T7S9SriE19doNALwqIHNxlWerFeE/edit#heading=h.z9wdvrobyao9.

Reply:
Is request for the KMS credentials really that frequent? I honestly don't know how often it is needed.

The token response includes expiry information that can be useful to maintain a cache of the token, if necessary.

Reply:
Potentially. libmongocrypt will cache the decrypted Data Encryption Key for 1 minute. So if (automatic|explicit) (encryption|decryption) has all keys cached, the `MONGOCRYPT_CTX_NEED_KMS_CREDENTIALS` state is not expected to be entered.

Other operations always need KMS. For example: [creating a data key](https://github.com/mongodb/libmongocrypt/blob/db6d3ab9a7f7d5ef20a95341a4007196c454b705/src/mongocrypt-ctx-datakey.c#L506) or rewrapping data keys.

Reply:
I forgot about this. If a user receives an error attempting to create > 5 data keys in one second, I think this caching the token in the driver may be necessary.